Fix db browser

kimakan opened the pull request:
This PR fixes two issues
 - the list of schemas and tables in the DB browser do not update after click on the column anymore. Also, the list of schemas is not updated after the click on a table.
 - The list of tables/columns now always start at the top of the list if the list is updated by clicking new schema/table.

jochenklar commented:
Hi @kimakan , both the blocks for `activeItem.type == 'table'` and  `activeItem.type == 'column'` are for the `/managament/` interface, when you create a new table or column, it should jump to this new item. However, this seems to conflict with the user schema (+ a few more bugs ...). I fixed the issues locally. 

The main problem was that the reduce search will find a wrong table, if the a tables have columns with the same name (which should happen all the time not only in user tables). The search is only needed if the item is new, otherwise for e.g. a table, the open schema does not be updated. So we can check for `activeItem.schema` or `activeItem.table` which is only available on new items in the management interface.

This should work:

```js
} else if (activeItem.type == 'table') {
        if (!isNil(activeItem.schema)) {
          // this is a newly created table, search for the schema and table
          const [schema, table] = schemas.reduce((result, schema) => {
            return schema.id == activeItem.schema ? (
              [schema, (schema.tables || []).find(t => isEqual(t, activeItem))]
            ) : result
          }, [])

          if (schema) {
            setOpenSchema(schema)
            setVisibleTables(schema.tables)
          }

          if (table) {
            setOpenTable(table)
            setVisibleColumns(table.columns)
          }
        } else {
          // this a regular click on a table, just open it
          setOpenTable(activeItem)
          setVisibleColumns(activeItem.columns)
        }
      } else if (activeItem.type == 'column') {
        if (!isNil(activeItem.table)) {
          // this is a newly created column, search for the schema and table
          const [schema, table] = schemas.reduce((result, schema) => {
            const table = (schema.tables || []).find(t => (t.id == activeItem.table))
            return isNil(table) ? result : [schema, table]
          }, [])

          if (schema) {
            setOpenSchema(schema)
            setVisibleTables(schema.tables)
          }

          if (table) {
            setOpenTable(table)
            setVisibleColumns(table.columns)
          }
```

Further down, there is another problem:

```js
                        {
                          openSchema && (isEqual(openSchema, schema)) && (
                            <div className="ms-auto"><i className="bi bi-chevron-right"></i></div>
                          )
                        }
```


```js
                        {
                          openTable && (isEqual(openTable, table)) && (
                            <div className="ms-auto"><i className="bi bi-chevron-right"></i></div>
                          )
                        }
```

We need `isEqual` here since the `userSchema` tables/columns don't have an id.

kimakan commented:
@jochenklar Thank you for the explanation. It makes perfectly sense. I've added your suggestions to the PR. In the mean time, I found some other bugs or things that were just annoying and fixed them.
 - Adding new table to the metadata was not working if the table didn't exist in the scientific database
 - Schema name or table name in the modal of the new item was always set to the first item in the list. Better, use the selected schema/table.

jochenklar commented on daiquiri/metadata/assets/js/components/Schemas.js:
```diff
@@ -44,40 +47,56 @@ const Schemas = ({ schemas, activeItem, setActiveItem, getTooltip, onDoubleClick
+            return schema.id == activeItem.schema ? (
+              [schema, (schema.tables || []).find(t => isEqual(t, activeItem))]
+            ) : result
+          }, [] )
```
Can you add a newline here?

kimakan replied:
Done!

jochenklar commented on daiquiri/metadata/assets/js/components/Schemas.js:
```diff
@@ -44,40 +47,56 @@ const Schemas = ({ schemas, activeItem, setActiveItem, getTooltip, onDoubleClick
         }
 
         setOpenTable(table)
+        if (refListTables.current) {
```
This scrolls the `dq-browser-list` for tables/columns to its top when it is set to a different schema/table, right? (I have to few items in my test setup.)

kimakan replied:
Yes, this scrolls the list of the tables to the top if a schema is clicked. The column list works similar.

jochenklar commented on daiquiri/metadata/models.py:
```diff
@@ -244,7 +244,11 @@ def admin_url(self):
+            self.type = metadata['type']
+        except KeyError:
+            # if the table does not exist in the database then do nothing
+            return
```
Here I would make `return None` explicit or maybe use `return {}`.

jochenklar replied:
Forget it, now I see that `discover` is not supposed to return anything.

jochenklar reviewed (approved):
Thanks!
